types/wlr_pointer_constraints_v1.c

```diff
@@ -0,0 +1,360 @@
+#include <stdbool.h>
+#include <stdlib.h>
+#include <pixman.h>
+#include <wayland-server.h>
+#include <wlr/types/wlr_box.h>
+#include <wlr/types/wlr_pointer_constraints_v1.h>
+#include <wlr/types/wlr_region.h>
+#include <wlr/util/log.h>
+
+#include "util/signal.h"
+
+static const struct zwp_locked_pointer_v1_interface locked_pointer_impl;
+static const struct zwp_confined_pointer_v1_interface confined_pointer_impl;
+static const struct zwp_pointer_constraints_v1_interface pointer_constraints_impl;
+static void pointer_constraint_destroy(struct wlr_pointer_constraint_v1 *constraint);
+
+static struct wlr_pointer_constraint_v1 *pointer_constraint_from_resource(
+		struct wl_resource *resource) {
+	assert(
+		wl_resource_instance_of(
+			resource, &zwp_confined_pointer_v1_interface,
+				&confined_pointer_impl) ||
+		wl_resource_instance_of(
+			resource, &zwp_locked_pointer_v1_interface,
+				&locked_pointer_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static struct wlr_pointer_constraints_v1 *pointer_constraints_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwp_pointer_constraints_v1_interface,
+		&pointer_constraints_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void resource_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static void pointer_constraint_destroy(struct wlr_pointer_constraint_v1 *constraint) {
+	if (constraint == NULL) {
+		return;
+	}
+
+	wlr_log(WLR_DEBUG, "destroying constraint %p", constraint);
+
+	wl_resource_set_user_data(constraint->resource, NULL);
+	wl_list_remove(&constraint->link);
+	wl_list_remove(&constraint->surface_commit.link);
+	wl_list_remove(&constraint->surface_destroy.link);
+	wl_list_remove(&constraint->seat_destroy.link);
+
+	wlr_signal_emit_safe(
+		&constraint->pointer_constraints->events.constraint_destroy,
+		constraint);
+
+	if (constraint->current.region) {
+		pixman_region32_fini(constraint->current.region);
+		free(constraint->current.region);
+	}
+	if (constraint->pending.region &&
+			constraint->current.region != constraint->pending.region) {
+		pixman_region32_fini(constraint->pending.region);
+		free(constraint->pending.region);
+	}
+	pixman_region32_fini(&constraint->region);
+	free(constraint);
+}
+
+static void pointer_constraint_destroy_resource(struct wl_resource *resource) {
+	struct wlr_pointer_constraint_v1 *constraint =
+		pointer_constraint_from_resource(resource);
+
+	pointer_constraint_destroy(constraint);
+}
+
+static void pointer_constraint_set_region(
+		struct wlr_pointer_constraint_v1 *constraint,
+		struct wl_resource *region_resource) {
+	if (constraint->pending.region &&
+			constraint->current.region != constraint->pending.region) {
+		pixman_region32_fini(constraint->pending.region);
+		free(constraint->pending.region);
+	}
+
+	if (region_resource) {
+		constraint->pending.region = calloc(1, sizeof(pixman_region32_t));
```

Comment:
No need to allocate regions. Just use empty regions instead of NULL.

Reply:
Won't the inner boxes be allocated anyways though? In addition committing changes will be more complex, and you will have to copy all the boxes on each commit, or am I missing something?

Reply:
I don't get what you mean by that

Reply:
Bump